guppylang/compiler/core.py

```diff
@@ -3,17 +3,55 @@
 from typing import cast
 
 from hugr import Wire, ops
-from hugr.build.dfg import DP, DfBase
+from hugr.build.dfg import DP, DefinitionBuilder, DfBase
 
 from guppylang.checker.core import FieldAccess, Place, PlaceId, Variable
-from guppylang.definition.common import CompiledDef, DefId
+from guppylang.definition.common import CheckedDef, CompilableDef, CompiledDef, DefId
 from guppylang.error import InternalGuppyError
 from guppylang.tys.ty import StructType
 
-CompiledGlobals = dict[DefId, CompiledDef]
 CompiledLocals = dict[PlaceId, Wire]
 
 
+class CompiledGlobals:
+    """Compilation context containing all available definitions.
+
+    This context drives the Hugr lowering by keeping track of which definitions are
+    used when lowering the definitions provided via the `required` keyword. The
```

Comment:
I dropped the `required` keyword, as we already have other ways to add to the worklist (and `compile_outer`) from outside (e.g. for nested functions) - `__getitem__` does basically the same thing